examples/deal.II/bps.h

```diff
@@ -0,0 +1,878 @@
+    CeedOperatorSetField(op_apply, "qdata", q_data_restriction, CEED_BASIS_COLLOCATED, q_data);
+    CeedOperatorSetField(op_apply, "v", sol_restriction, sol_basis, CEED_VECTOR_ACTIVE);
+  }
+
+  /**
+   * Perform matrix-vector product.
+   */
+  void
+  vmult(VectorType &dst, const VectorType &src) const override
+  {
+    // communicate: update ghost values
+    src.update_ghost_values();
+
+    if (dof_handler.get_fe().n_components() == 1)
+      {
+        // create libCEED view on deal.II vectors
+        VectorTypeCeed src_ceed(ceed, src);
+        VectorTypeCeed dst_ceed(ceed, dst);
+
+        // apply operator
+        CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+      }
+    else // TODO: needed for multiple components
+      {
+        // allocate space for block vectors
+        src_tmp.reinit(this->extended_local_size(), true);
+        dst_tmp.reinit(this->extended_local_size(), true);
+
+        // create libCEED view on deal.II vectors
+        VectorTypeCeed src_ceed(ceed, src_tmp);
+        VectorTypeCeed dst_ceed(ceed, dst_tmp);
+
+        copy_to_block_vector(src_tmp, src); // copy to block vector
+
+        // apply operator
+        CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+
+        copy_from_block_vector(dst, dst_tmp); // copy from block vector
+      }
+
+    // communicate: compress
+    src.zero_out_ghost_values();
+    dst.compress(VectorOperation::add);
+
+    // apply constraints: we assume homogeneous DBC
+    constraints.set_zero(dst);
+  }
+
+  /**
+   * Initialized vector.
+   */
+  void
+  initialize_dof_vector(VectorType &vec) const override
+  {
+    vec.reinit(partitioner);
+  }
+
+  /**
+   * Compute inverse of diagonal.
+   */
+  void
+  compute_inverse_diagonal(VectorType &diagonal) const override
+  {
+    this->initialize_dof_vector(diagonal);
+
+    VectorTypeCeed diagonal_ceed(ceed, diagonal);
+
+    CeedOperatorLinearAssembleDiagonal(op_apply, diagonal_ceed(), CEED_REQUEST_IMMEDIATE);
+
+    const unsigned int n_components = dof_handler.get_fe().n_components();
+
+    if (n_components > 1) // TODO: needed for multiple components
+      {
+        VectorType tmp(diagonal);
+
+        copy_from_block_vector(tmp, diagonal);
+
+        std::swap(tmp, diagonal);
+      }
+
+    diagonal.compress(VectorOperation::add);
+
+    for (auto &i : diagonal)
+      i = (std::abs(i) > 1.0e-10) ? (1.0 / i) : 1.0;
+  }
+
+private:
+  /**
+   * Wrapper around a deal.II vector to create a libCEED vector view.
+   */
+  class VectorTypeCeed
+  {
+  public:
+    /**
+     * Constructor.
+     */
+    VectorTypeCeed(const Ceed &ceed, const VectorType &vec)
+    {
+      const unsigned int n_dofs =
+        vec.get_partitioner()->locally_owned_size() + vec.get_partitioner()->n_ghost_indices();
+
+      CeedVectorCreate(ceed, n_dofs, &vec_ceed);
+      CeedVectorSetArray(vec_ceed, CEED_MEM_HOST, CEED_USE_POINTER, vec.get_values());
```

Comment:
I think this is be the problem I founding using the memcheck backend, which would also be an issue for all GPU backends. In our GPU backends (which the memcheck backend emulates), each CeedVector has two memory spaces, `CEED_MEM_HOST` and `CEED_MEM_DEVICE`. The CeedVector manages synchronization between these two spaces, but only if the caller uses `CeedVectorSetArray` and `CeedVectorTakeArray` in pairs.

Here's how we do that in the PETSc examples every time `CeedOperatorApply` is called
```C
  // Setup CEED vectors
  PetscCall(VecGetArrayReadAndMemType(user->X_loc, (const PetscScalar **)&x, &x_mem_type));
  PetscCall(VecGetArrayAndMemType(user->Y_loc, &y, &y_mem_type));
  CeedVectorSetArray(user->x_ceed, MemTypeP2C(x_mem_type), CEED_USE_POINTER, x);
  CeedVectorSetArray(user->y_ceed, MemTypeP2C(y_mem_type), CEED_USE_POINTER, y);

  // Apply CEED operator
  CeedOperatorApply(user->op, user->x_ceed, user->y_ceed, CEED_REQUEST_IMMEDIATE);

  // Restore PETSc vectors
  CeedVectorTakeArray(user->x_ceed, MemTypeP2C(x_mem_type), NULL);
  CeedVectorTakeArray(user->y_ceed, MemTypeP2C(y_mem_type), NULL);
  PetscCall(VecRestoreArrayReadAndMemType(user->X_loc, (const PetscScalar **)&x));
  PetscCall(VecRestoreArrayAndMemType(user->Y_loc, &y));
```

I'm not sure where best to put that logic though. Just adding
```C
    /**
     * Destructor: destroy vector view.
     */
    ~VectorTypeCeed()
    {
      CeedScalar *ptr;
      CeedVectorTakeArray(vec_ceed, CEED_MEM_HOST, &ptr);
      CeedVectorDestroy(&vec_ceed);
    }
```
isn't enough